src/api/index.tsx

```diff
@@ -0,0 +1,64 @@
+import { ApiState } from '../reducers/api';
+
+export enum HttpMethod {
+  DELETE = 'DELETE',
+  GET = 'GET',
+  PATCH = 'PATCH',
+  POST = 'POST',
+  PUT = 'PUT',
+}
+
+type CallApiParams = {
+  endpoint: string;
+  method?: HttpMethod;
+  credentials?: boolean;
+  version?: string;
+  apiState: ApiState;
+};
+
+type Headers = {
+  [name: string]: string;
+};
+
+export const callApi = async ({
+  apiState,
+  endpoint,
+  credentials = false,
+  method = HttpMethod.GET,
+  version = 'v4',
+}: CallApiParams): Promise<object | { error: Error }> => {
+  if (!endpoint.startsWith('/')) {
+    endpoint = `/${endpoint}`;
+  }
+  if (!endpoint.endsWith('/')) {
+    endpoint = `${endpoint}/`;
+  }
+
+  const headers = {} as Headers;
+  if (apiState.authToken) {
+    headers['Authorization'] = `Bearer ${apiState.authToken}`;
+  }
+
+  try {
+    const response = await fetch(`/api/${version}${endpoint}`, {
+      method,
+      headers,
+      credentials: credentials ? 'include' : undefined,
+    });
+
+    return await response.json();
```

Comment:
Your patch shows a different implementation than what you wrote and it is more aligned with my previous comment:

> [...] unless we improve this function to return an error object when the response is not OK.

```diff
+    throw new Error(`Unexpected status for ${method} ${endpoint}: ${response.status}`);
```

Why not returning this error instead?

This would allow callers to write:

```js
const data = await callApi({ endpoint: '/foo' });

if (!data.error) {
  // happy path
} else {
  // error handling, like showing the error to the user
}
```

It is much more concise and throwing in async functions always lead to leaky promises (and this famous "node unhandled Promise Rrjections" warnings).

We could log a warning when `!response.ok` so that we get the best of both worlds: a nice API and developer feedback if something went wrong.

Reply:
> It is much more concise and throwing in async functions always lead to leaky promises (and this famous "node unhandled Promise Rrjections" warnings).
> 
> We could log a warning when `!response.ok` so that we get the best of both worlds: a nice API and developer feedback if something went wrong.

I've implemented this suggestion to show what I was thinking